addons/mail/static/src/discuss/call/common/rtc_service.js

```diff
@@ -127,6 +128,35 @@ export class Rtc extends Record {
     /** @type {import("@mail/static/libs/odoo_sfu/odoo_sfu").SfuClient} */
     sfuClient = undefined;
 
+    lastActions = {};
+    actionsStack = [];
+    lastSelfCallAction = undefined;
+
+    callActions = Record.attr([], {
+        compute() {
+            return callActionsRegistry
+                .getEntries()
+                .map(([key, action]) => [key, action.isActive(this.selfSession)]);
+        },
+        onUpdate() {
+            for (const [key, isActive] of this.callActions) {
+                if (isActive !== this.lastActions[key]) {
+                    if (isActive) {
+                        if (!this.actionsStack.includes(key)) {
+                            this.actionsStack.unshift(key);
+                        }
+                    } else {
+                        this.actionsStack.splice(this.actionsStack.indexOf(key), 1);
+                    }
+                }
+            }
+
+            this.lastSelfCallAction = this.actionsStack[0];
+
+            this.lastActions = Object.fromEntries(this.callActions);
+        },
+    });
```

Comment:
Note that if each action was a record, `callActions` would simply be a `Record.many("CallAction")` that is computed once, after items are added in registry.

`actionsStack` would also be a `Record.many("CallAction")` that simply has a `Record.one()` inverse on the action. This would put the responsibility of each action to insert itself in the `actionsStack`. In theory, this means a change of active action results in `O(1)` operation, instead of current implementation where you have to parse all call actions all the time so `O(n)`.

`lastSelfCallAction` would simply become a computed `Record.one("CallAction")` to `actionsStack[0]`.

I think current implementation is fine, though it could be inefficient and result to stuttering in theory. Also the fields lack typing, defining models allow for easy typing by just naming the model.

Reply:
You can omit this comment. Though this is interesting, I think the simplicity of current code is good enough for now.